File: 0038-count-and-say/0038-count-and-say.java
class Solution {
    private String[] result = new String[31];

    public Solution() {
        preCalculate();
    }

    public String countAndSay(int n) {
        return result[n];
    }

    private void preCalculate() {
        result[0] = "0";
        result[1] = "1";

        for (int i = 2; i <= 30; i++) {
            char previous;
            int count = 1;
            String now = "";
            String lastWord = result[i - 1];

            previous = lastWord.charAt(0);

            for (int j = 1; j < lastWord.length(); j++) {
                if (lastWord.charAt(j) == previous) {
                    count++;
                } else {
                    now += (String.valueOf(count) + String.valueOf(previous));
                    previous = lastWord.charAt(j);
                    count = 1;
                }
            }

            now += (String.valueOf(count) + String.valueOf(previous));

            result[i] = now;
        }
    }
}

File: 0038-count-and-say/0038-count-and-say.py
class Solution:
    def countAndSay(self, n: int) -> str:
        if n == 1:
            return "1"
        
        return self.generateRLE(self.countAndSay(n - 1))


    def generateRLE(self, number: str) -> str:
        rle = ""
        count = 1
        digit = number[0]

        for i in range(1, len(number)):
            if number[i] == number[i - 1]:
                count += 1
            else:
                rle += f"{count}{digit}"
                count = 1
                digit = number[i]

        rle += f"{count}{digit}"
        return rle
